### ISSUE TITLE
hyperparameter tuning

### ISSUE THREAD

**Issue by maekou-AI** [maekou-AI/cmi_piu#4]
ハイパーパラメータのチューニングを簡易的に行う

**Comment by maekou-AI** [comment]
learning_rate: 0.01 ~ 1.0
max_depth: 1 ~ 16
num_leaves: 2 ~ 2^max_depth
min_data_in_leaf: 0-300
subsample: 0.01-1.0
feature_fraction: 0.1-1.0
reg_lambda: 1e-9 ~ 100
reg_alpha: 0-100

**Comment by maekou-AI** [comment]
def train_kfold(features, targets, categorical_features, n_splits, base_params, weights=None):
    if weights is None:
        weights = np.ones(len(features))

    skf = StratifiedKFold(
        n_splits=n_splits, shuffle=True, random_state=1
    )
    boosters = []
    val_qwks = []
    oof_preds = np.zeros(len(features))

    for i, (trn_idx, val_idx) in enumerate(skf.split(features, targets)):
        trn_data = lgb.Dataset(
            data=features.iloc[trn_idx],
            label=targets.iloc[trn_idx],
            weight=weights[trn_idx],
            categorical_feature=categorical_features,
            free_raw_data=False
        )
        val_data = lgb.Dataset(
            data=features.iloc[val_idx],
            label=targets.iloc[val_idx],
            weight=weights[val_idx],
            categorical_feature=categorical_features,
            free_raw_data=False
        )

        if i == 0:          
            def objective(trial):
                params = {
                    # "learning_rate": trial.suggest_float("learning_rate", 0.01, 1.0),
                    "num_leaves": trial.suggest_int("num_leaves", 2, 256),
                    "max_depth": trial.suggest_int("max_depth", 1, 16),
                    "min_data_in_leaf": trial.suggest_int("min_data_in_leaf", 0, 300),
                    "bagging_fraction": trial.suggest_uniform("bagging_fraction", 0.4, 1.0),
                    "bagging_freq": trial.suggest_int("bagging_freq", 1, 7),
                    "feature_fraction": trial.suggest_uniform("feature_fraction", 0.4, 1.0),
                    "min_child_samples": trial.suggest_int("min_child_sample", 5, 100),
                    "lambda_l1": trial.suggest_loguniform("lambda_l1", 1e-8, 10.0),
                    "lambda_l2": trial.suggest_loguniform("lambda_l2", 1e-8, 10.0),
                }
                eval_result = {}
                callbacks = [
                    lgb.early_stopping(STOPPING_ROUNDS),
                    lgb.log_evaluation(VERBOSE_EVAL),
                    lgb.record_evaluation(eval_result),
                ]
                booster = lgb.train(
                    {**base_params, **params},
                    trn_data,
                    valid_sets=[trn_data, val_data],
                    valid_names=["trn", "val"],
                    num_boost_round=NUM_BOOST_ROUND,
                    feval=quadratic_weighted_kappa,
                    callbacks=callbacks
                )
                return eval_result["val"]["QWK"][booster.best_iteration - 1]
            
            study = optuna.create_study(direction="maximize")
            study.optimize(objective, n_trials=100)
            best_params = study.best_params
            print("Best hyperparameters:", best_params)
        break
